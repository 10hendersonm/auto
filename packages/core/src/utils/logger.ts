import signale, { Signale } from 'signale';

export interface ILogger {
  /** The level at which to log messages */
  logLevel: LogLevel;
  /** The default logger. Always on */
  log: signale.Signale<signale.DefaultMethods>;
  /** The verbose log. Has more debug logs */
  verbose: signale.Signale<signale.DefaultMethods>;
  /** The very verbose log. Has all debug logs */
  veryVerbose: signale.Signale<signale.DefaultMethods>;
}

/** Create a dummy logger for testing. */
export function dummyLog(): ILogger {
  return {
    logLevel: undefined,
    log: new Signale({ disabled: true }),
    verbose: new Signale({ disabled: true }),
    veryVerbose: new Signale({ disabled: true })
  };
}

export type LogLevel = 'verbose' | 'veryVerbose' | undefined;

/** Create a logger the the given log level. */
export default function createLog(mode: LogLevel): ILogger {
  return {
    logLevel: mode,
    log: new Signale(),
    verbose: new Signale({
      disabled: mode !== 'verbose' && mode !== 'veryVerbose'
    }),
    veryVerbose: new Signale({
      disabled: mode !== 'veryVerbose'
    })
  };
}
