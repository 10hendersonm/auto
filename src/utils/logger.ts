import signale, { Signale } from 'signale';

export interface ILogger {
  log: signale.Signale<signale.DefaultMethods>;
  verbose: signale.Signale<signale.DefaultMethods>;
  veryVerbose: signale.Signale<signale.DefaultMethods>;
}

export function dummyLog(): ILogger {
  return {
    log: new Signale({ disabled: true }),
    verbose: new Signale({ disabled: true }),
    veryVerbose: new Signale({ disabled: true })
  };
}

export type LogLevel = 'verbose' | 'veryVerbose' | undefined;

export default function createLog(mode: LogLevel): ILogger {
  return {
    log: new Signale(),
    verbose: new Signale({
      disabled: mode !== 'verbose' && mode !== 'veryVerbose'
    }),
    veryVerbose: new Signale({
      disabled: mode !== 'veryVerbose'
    })
  };
}
