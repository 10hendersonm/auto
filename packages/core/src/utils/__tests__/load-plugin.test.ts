import path from 'path';
import loadPlugin from '../load-plugins';
import { dummyLog } from '../logger';

const logger = dummyLog();

jest.mock(
  'auto-plugin-foo',
  () => ({
    default: class {
      name = 'foo';
    }
  }),
  {
    virtual: true
  }
);
jest.mock(
  '@my-scope/auto-plugin-bar',
  () => ({
    default: class {
      name = 'bar';
    }
  }),
  { virtual: true }
);
jest.mock(
  '@auto-it/baz',
  () => ({
    default: class {
      name = 'baz';
    }
  }),
  {
    virtual: true
  }
);

describe('loadPlugins', () => {
  test('should load official plugins', () => {
    expect(loadPlugin(['baz', {}], logger)?.name).toBe('baz');
    expect(loadPlugin(['@auto-it/baz', {}], logger)?.name).toBe('baz');
  });

  test('should load community plugins', () => {
    expect(loadPlugin(['foo', {}], logger)?.name).toBe('foo');
    expect(loadPlugin(['auto-plugin-foo', {}], logger)?.name).toBe('foo');
  });

  test('should load scoped plugins', () => {
    expect(loadPlugin(['@my-scope/auto-plugin-bar', {}], logger)?.name).toBe(
      'bar'
    );
  });

  test('should require custom plugins -- fallback to cwd', () => {
    expect(
      loadPlugin([path.join(__dirname, './test-plugin.ts'), {}], logger)
    ).toStrictEqual(
      expect.objectContaining({
        name: 'foo',
        config: {}
      })
    );
  });

  test('should require custom plugins -- surface errors', () => {
    expect(() =>
      loadPlugin(
        [path.join(__dirname, './test-plugin-malformed.js'), {}],
        logger
      )
    ).toThrow();
  });

  test('should load config', () => {
    expect(
      loadPlugin(
        [path.join(__dirname, './test-plugin.ts'), 'do the thing'],
        logger
      )
    ).toStrictEqual(
      expect.objectContaining({
        name: 'foo',
        config: 'do the thing'
      })
    );
  });
});
