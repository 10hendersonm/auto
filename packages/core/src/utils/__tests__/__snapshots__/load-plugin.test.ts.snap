// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`listPlugins should get plugins from autorc 1`] = `
Array [
  "Found the following plugins in your .autorc:

- npm (@auto-it/npm)",
]
`;

exports[`listPlugins should get plugins from local module 1`] = `
Array [
  "Found the following plugins in your .autorc:

- npm (@auto-it/npm)",
]
`;

exports[`listPlugins should get plugins from local module 2`] = `
Array [
  "Found the following plugins installed in your project:

- npm (../../../../some/folder/node_modules/@auto-it/npm)
- released (../../../../some/folder/node_modules/@auto-it/released)",
]
`;

exports[`listPlugins should get plugins from local module 3`] = `
Array [
  "Found the following plugins globally installed in your environment:

- npm (/some/folder/node_modules/@auto-it/npm)
- released (/some/folder/node_modules/@auto-it/released)",
]
`;
